### PR TITLE
feat: add scheme start state

### PR DESCRIPTION
**Pull request by anonrig**

### CODE REVIEW

File: include/ada/url.h [hunk 1]
```diff
@@ -37,23 +45,45 @@ namespace ada {
...
+     * User agents, especially conformance checkers, are encouraged to report them somewhere.
+     */
+    bool has_validation_error = false;
+
+    /**
+     * A URL includes credentials if its username or password is not the empty string.
+     * @return
+     */
+    bool includes_credentials() const {
+      return username != nullptr || password != nullptr;
```
Comment (lemire) [inline include/ada/url.h:77]: This is not going to be correct if usename (e.g.) is a string view. I think you want something like !username.empty().

**Comment by lemire** [comment]
In C++11, you should not need to do the following assignments in the definition of a class. One gets default initialization by default which in the case of a `bool` variable, sets it to false automagically.

```
    bool at_sign_seen = false;
    bool inside_brackets = false;
    bool password_token_seen = false;
```

**Comment by lemire** [comment]
Sorry. Above is wrong because you roll your own constructor. Let me check.

File: include/ada/url.h (hunk 1 above)
Comment (anonrig) [reply include/ada/url.h:77]: You're right. Thanks.

**Comment by lemire** [comment]
Random comments which you *may* consider...

1. There is currently a mix of different styles. The most common in C++ (at least in the standard library) is all lower-case (except for macros) with underscores: `void parse_state()`. Whatever one does, we should not mix `parseState()` and `parse_state()` in the same code base.
2. As much as possible, one should try to initialize the variables in the following manner (for example):   ```
```C++
parser::parser(const std::string_view input, std::optional<ada::url> optional_base_url, 
   std::optional<ada::encoding_type> encoding_override):  buffer{}, 
   pointer{input.begin()}, encoding{encoding_override.value_or(ada::encoding_type::UTF8)} {
...
}
``` 
That is, once you enter the curly bracket part of the constructor, your attributes should have a clear state. I think it is considered form to do the initialization in the curly-bracket component of the constructor.
3. In the header file, you can make sure the attributes of your class are always initialized with the curly-bracket trick:
```
    std::string_view input{};
    std::string buffer{};
    bool at_sign_seen{}; // will be false
    bool inside_brackets{};
    bool password_token_seen{};
```
3. Tools like clang-format will allow one to set a line limit. I think it is often considered bad form to have lines much longer than 80 characters.

**Comment by lemire** [comment]
**These are not requests for changes.*** They are comments.

**Comment by anonrig** [comment]
Thank you for the review @lemire. Much appreciated. Let's merge this now, and I'll apply your comments in a separate pull request.

**Pull request merged**